File: bindings/swift/Sources/Proven/SafeUnit.swift
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/// Physical unit conversions delegated to libproven FFI.
///
/// Length and temperature conversion via the formally verified
/// Idris 2 core.

import CProven

/// Length units supported by SafeUnit.
public enum LengthUnit: UInt32, Sendable {
    case meters = 0
    case kilometers = 1
    case centimeters = 2
    case millimeters = 3
    case feet = 4
    case inches = 5
    case miles = 6
    case yards = 7
}

/// Temperature units supported by SafeUnit.
public enum TempUnit: UInt32, Sendable {
    case celsius = 0
    case fahrenheit = 1
    case kelvin = 2
}

public enum SafeUnit {
    /// Convert length between units.
    public static func convertLength(_ value: Double, from: LengthUnit, to: LengthUnit) -> Result<Double, ProvenError> {
        let result = proven_unit_convert_length(value, ProvenLengthUnit(rawValue: from.rawValue), ProvenLengthUnit(rawValue: to.rawValue))
        if let error = ProvenError.fromStatus(result.status) {
            return .failure(error)
        }
        return .success(result.value)
    }

    /// Convert temperature between units.
    public static func convertTemp(_ value: Double, from: TempUnit, to: TempUnit) -> Result<Double, ProvenError> {
        let result = proven_unit_convert_temp(value, ProvenTempUnit(rawValue: from.rawValue), ProvenTempUnit(rawValue: to.rawValue))
        if let error = ProvenError.fromStatus(result.status) {
            return .failure(error)
        }
        return .success(result.value)
    }
}

File: bindings/deno/src/safe_unit.ts
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

/**
 * SafeUnit - Physical unit conversions.
 *
 * Thin FFI wrapper: all computation delegates to libproven (Idris 2 + Zig).
 *
 * @module
 */

import { getLib, ProvenStatus, statusToError } from './ffi.ts';
import { err, ok, type Result } from './result.ts';

/**
 * Length unit enumeration (matches Zig LengthUnit enum).
 */
export const LengthUnit = {
  METERS: 0,
  KILOMETERS: 1,
  CENTIMETERS: 2,
  MILLIMETERS: 3,
  FEET: 4,
  INCHES: 5,
  MILES: 6,
  YARDS: 7,
} as const;

/**
 * Temperature unit enumeration (matches Zig TempUnit enum).
 */
export const TemperatureUnit = {
  CELSIUS: 0,
  FAHRENHEIT: 1,
  KELVIN: 2,
} as const;

/**
 * Safe unit conversion operations backed by formally verified Idris 2 code.
 */
export class SafeUnit {
  /**
   * Convert a length value between units.
   *
   * @param value - The value to convert.
   * @param from - Source LengthUnit.
   * @param to - Target LengthUnit.
   * @returns Result containing the converted value or an error string.
   */
  static convertLength(value: number, from: number, to: number): Result<number> {
    const symbols = getLib();
    const result = symbols.proven_unit_convert_length(value, from, to);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }

  /**
   * Convert a temperature value between units.
   *
   * @param value - The value to convert.
   * @param from - Source TemperatureUnit.
   * @param to - Target TemperatureUnit.
   * @returns Result containing the converted value or an error string.
   */
  static convertTemp(value: number, from: number, to: number): Result<number> {
    const symbols = getLib();
    const result = symbols.proven_unit_convert_temp(value, from, to);
    if (result[0] !== ProvenStatus.OK) return err(statusToError(result[0]));
    return ok(result[1]);
  }
}
